src/device-manager/java/src/nl/Weave/DataManagement/GenericTraitUpdatableDataSinkImpl.java

```diff
@@ -0,0 +1,444 @@
+    }
+
+    @Override
+    public void set(String path, byte[] value)
+    {
+        set(path, value, false);
+    }
+
+    @Override
+    public void setNULL(String path)
+    {
+        setNULL(path, false);
+    }
+
+    @Override
+    public int getInt(String path)
+    {
+        if (mTraitInstancePtr == 0)
+        {
+            Log.e(TAG, "unexpected err, mTraitInstancePtr is 0");
+            return 0;
+        }
+        return (int)getInt(mTraitInstancePtr, path);
+    }
+
+    @Override
+    public long getLong(String path)
+    {
+        if (mTraitInstancePtr == 0)
+        {
+            Log.e(TAG, "unexpected err, mTraitInstancePtr is 0");
+            return 0;
+        }
+        return (long)getInt(mTraitInstancePtr, path);
+    }
+
+    @Override
+    public BigInteger getBigInteger(String path, int bitLen)
+    {
+        long value = (long)getInt(mTraitInstancePtr, path);
+        String valStr = Long.toString(value);
+        BigInteger bigInt = new BigInteger(valStr);
+        if (mTraitInstancePtr == 0)
+        {
+            Log.e(TAG, "unexpected err, mTraitInstancePtr is 0");
+            return bigInt;
+        }
+        if (bitLen != 8 && bitLen != 16 && bitLen != 32 && bitLen != 64)
+        {
+            Log.e(TAG, "Not support bit len, return original value" + bitLen);
+            return bigInt;
+        }
+        BigInteger twoComplement = BigInteger.ONE.shiftLeft(bitLen);
+
+        if (bigInt.compareTo(BigInteger.ZERO) < 0)
+            bigInt = bigInt.add(twoComplement);
+        if (bigInt.compareTo(twoComplement) >= 0)
+        {
+            Log.e(TAG, "overflow range:" + bitLen);
+            return twoComplement;
+        }
+        else if (bigInt.compareTo(BigInteger.ZERO) < 0 )
+        {
+            Log.e(TAG, "incorrect range:" + bitLen);
+            return BigInteger.ZERO;
+        }
+        else
+        {
+            return bigInt;
+        }
+    }
+
+    @Override
+    public double getDouble(String path)
+    {
+        if (mTraitInstancePtr == 0)
+        {
+            Log.e(TAG, "unexpected err, mTraitInstancePtr is 0");
+            return 0;
+        }
+        return getDouble(mTraitInstancePtr, path);
+    }
+
+    @Override
+    public boolean getBoolean(String path)
+    {
+        if (mTraitInstancePtr == 0)
+        {
+            Log.e(TAG, "unexpected err, mTraitInstancePtr is 0");
+            return false;
+        }
+        return getBoolean(mTraitInstancePtr, path);
+    }
+
+    @Override
+    public String getString(String path)
+    {
+        if (mTraitInstancePtr == 0)
+        {
+            Log.e(TAG, "unexpected err, mTraitInstancePtr is 0");
+            return null;
+        }
+        return getString(mTraitInstancePtr, path);
+    }
+
+    @Override
+    public byte[] getBytes(String path)
+    {
+        if (mTraitInstancePtr == 0)
+        {
+            Log.e(TAG, "unexpected err, mTraitInstancePtr is 0");
+            return null;
```

Comment:
Would be preferable for these getters to throw the same IllegalStateException if closed, rather than return null

Reply:
Done, thanks